daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -689,6 +669,37 @@ private[lf] final class Compiler(
       translateExp(env.bindExprVar(tmpl.param, tmplArgPos), tmpl.observers)
     }
 
+  private[this] def compileToCachedContract(
+      tmplId: Identifier,
+      tmpl: Template,
+  ): (t.SDefinitionRef, SDefinition) =
+    topLevelFunction2(t.ToCachedContractDefRef(tmplId)) { (tmplArgPos, mbKeyPos, env) =>
+      SBuildCachedContract(
+        s.SEValue(STypeRep(TTyCon(tmplId))),
+        env.toSEVar(tmplArgPos),
+        t.SignatoriesDefRef(tmplId)(env.toSEVar(tmplArgPos)),
+        t.ObserversDefRef(tmplId)(env.toSEVar(tmplArgPos)),
+        tmpl.key match {
+          case None =>
+            s.SEValue.None
+          case Some(tmplKey) =>
+            s.SECase(
+              env.toSEVar(mbKeyPos),
+              List(
+                s.SCaseAlt(
```

Comment:
This pattern match seems completely redundant. We know that it will never hit the None case but we need to extract it due to the way things are setup here. I think before we were somewhat more clever by handling this within the builtin. Probably not worth worrying about but I think it’s maybe useful to run benchmarks to make sure we don’t regress.

I’d also suggest to add a comment that explains that this is redundant.

Reply:
> We know that it will never hit the None 

I think you are mistaken. For template with key, this new definition is used for `create`, normal `fetch`, `fetch-by-key` and `lookup-by-key`. In the two former cases `mbKeyPos` is `None`.

 > I think before we were somewhat more clever

Reply:
Oh I completely misread this. I thought `tmpl.key` cannot be Some if `mbKeyPos` is not also Some but that’s ofc complete BS: For things like exercise I can exercise by cid even if the template has a key. my bad